File: tools/zweitfaktor/totp.mjs
/**
 * DER CODE-RECHNER FÜR NODE (P5c/AP5, E-P5c-43). Einer je Sprache — die
 * anderen sind `totp.php` und `totp.py` daneben; alle drei rechnen dasselbe
 * und teilen denselben Zähler. Warum es den Zähler gibt und wie er sperrt,
 * steht im Kopf von `totp.php`.
 *
 * Anlass: F-P5c-33.
 *
 *   import { naechsterCode } from '../zweitfaktor/totp.mjs';
 *   const code = await naechsterCode();          // NADOKU_TOTP oder Sandbox
 *   const code = await naechsterCode(base32);    // ausdrücklich
 */
import { createHmac, createHash } from 'node:crypto';
import { openSync, closeSync, unlinkSync, statSync, readFileSync, writeFileSync, existsSync } from 'node:fs';

export const SANDBOX = 'PRUEFSTANDZWEITFAKTORNADOKU23456';

export function geheimnis() {
  const e = (process.env.NADOKU_TOTP || '').trim();
  return e !== '' ? e : SANDBOX;
}

function norm(b32) { return String(b32).replace(/[\s=]+/g, '').toUpperCase(); }

export function base32(b32) {
  const a = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';
  let bits = '';
  for (const z of norm(b32)) {
    const i = a.indexOf(z);
    if (i < 0) throw new Error(`Kein Base32: ${z}`);
    bits += i.toString(2).padStart(5, '0');
  }
  const out = [];
  for (let i = 0; i + 8 <= bits.length; i += 8) out.push(parseInt(bits.slice(i, i + 8), 2));
  return Buffer.from(out);
}

export function code(roh, schritt) {
  const b = Buffer.alloc(8);
  b.writeBigUInt64BE(BigInt(schritt));
  const h = createHmac('sha1', roh).update(b).digest();
  const o = h[19] & 0x0f;
  const n = (((h[o] & 0x7f) << 24) | (h[o + 1] << 16) | (h[o + 2] << 8) | h[o + 3]) % 1000000;
  return String(n).padStart(6, '0');
}

export function zaehlerdatei(b32) {
  const tmp = (process.env.TMPDIR || '/tmp').replace(/\/$/, '');
  const kennung = createHash('sha256').update(norm(b32)).digest('hex').slice(0, 12);
  return `${tmp}/nadoku-totp-${kennung}.schritt`;
}

const schlafe = (ms) => new Promise((r) => setTimeout(r, ms));

export async function naechsterCode(b32 = geheimnis()) {
  const datei = zaehlerdatei(b32);
  const sperre = `${datei}.sperre`;
  const bis = Date.now() + 60000;
  for (;;) {
    try { closeSync(openSync(sperre, 'wx')); break; } catch {
      try { if (Date.now() - statSync(sperre).mtimeMs > 10000) { unlinkSync(sperre); continue; } } catch {}
      if (Date.now() > bis) throw new Error(`Zählersperre hängt: ${sperre}`);
      await schlafe(50);
    }
  }
  try {
    const letzter = existsSync(datei) ? parseInt(readFileSync(datei, 'utf8').trim(), 10) || 0 : 0;
    let kandidat;
    for (;;) {
      const jetzt = Math.floor(Date.now() / 30000);
      kandidat = Math.max(letzter + 1, jetzt);
      if (kandidat <= jetzt + 1) break;
      await schlafe((kandidat - 1) * 30000 - Date.now() + 200);
    }
    writeFileSync(datei, String(kandidat));
    return code(base32(b32), kandidat);
  } finally {
    try { unlinkSync(sperre); } catch {}
  }
}

File: tools/zweitfaktor/totp.py
"""DER CODE-RECHNER FÜR PYTHON (P5c/AP5, E-P5c-43). Einer je Sprache — die
anderen sind `totp.php` und `totp.mjs` daneben; alle drei rechnen dasselbe
und teilen denselben Zähler. Warum es den Zähler gibt und wie er sperrt,
steht im Kopf von `totp.php`.

Anlass: F-P5c-33.

    sys.path.insert(0, str(WURZEL / 'tools' / 'zweitfaktor'))
    from totp import naechster_code
    code = naechster_code()          # NADOKU_TOTP oder Sandbox
    code = naechster_code(base32)    # ausdrücklich

Von der Kommandozeile: `python3 tools/zweitfaktor/totp.py` gibt den
nächsten Code aus.
"""
from __future__ import annotations

import base64
import hashlib
import hmac
import os
import struct
import time

SANDBOX = 'PRUEFSTANDZWEITFAKTORNADOKU23456'


def geheimnis() -> str:
    e = (os.environ.get('NADOKU_TOTP') or '').strip()
    return e or SANDBOX


def _norm(b32: str) -> str:
    return ''.join(b32.split()).replace('=', '').upper()


def roh(b32: str) -> bytes:
    n = _norm(b32)
    return base64.b32decode(n + '=' * (-len(n) % 8))


def code(geheim: bytes, schritt: int) -> str:
    h = hmac.new(geheim, struct.pack('>Q', schritt), hashlib.sha1).digest()
    o = h[19] & 0x0F
    n = (int.from_bytes(h[o:o + 4], 'big') & 0x7FFFFFFF) % 1_000_000
    return f'{n:06d}'


def zaehlerdatei(b32: str) -> str:
    tmp = (os.environ.get('TMPDIR') or '/tmp').rstrip('/')
    kennung = hashlib.sha256(_norm(b32).encode()).hexdigest()[:12]
    return f'{tmp}/nadoku-totp-{kennung}.schritt'


def naechster_code(b32: str | None = None) -> str:
    b32 = b32 or geheimnis()
    datei = zaehlerdatei(b32)
    sperre = datei + '.sperre'
    bis = time.time() + 60
    while True:
        try:
            os.close(os.open(sperre, os.O_CREAT | os.O_EXCL | os.O_WRONLY))
            break
        except FileExistsError:
            try:
                if time.time() - os.stat(sperre).st_mtime > 10:
                    os.unlink(sperre)
                    continue
            except FileNotFoundError:
                continue
            if time.time() > bis:
                raise RuntimeError(f'Zählersperre hängt: {sperre}')
            time.sleep(0.05)
    try:
        try:
            with open(datei, encoding='utf-8') as f:
                letzter = int(f.read().strip() or 0)
        except (FileNotFoundError, ValueError):
            letzter = 0
        while True:
            jetzt = int(time.time() // 30)
            kandidat = max(letzter + 1, jetzt)
            if kandidat <= jetzt + 1:
                break
            time.sleep((kandidat - 1) * 30 - time.time() + 0.2)
        with open(datei, 'w', encoding='utf-8') as f:
            f.write(str(kandidat))
        return code(roh(b32), kandidat)
    finally:
        try:
            os.unlink(sperre)
        except FileNotFoundError:
            pass


if __name__ == '__main__':
    print(naechster_code())
